lineapy/graph_reader/program_slice.py

```diff
@@ -3,19 +3,47 @@
 from typing import DefaultDict, List, Set
 
 from lineapy.data.graph import Graph
-from lineapy.data.types import ImportNode, LineaID, SourceCode
+from lineapy.data.types import CallNode, ImportNode, LineaID, SourceCode
 from lineapy.db.db import RelationalLineaDB
 from lineapy.utils.utils import prettify
 
 logger = logging.getLogger(__name__)
 
 
-def get_slice_graph(graph: Graph, sinks: List[LineaID]) -> Graph:
+def get_slice_graph(
+    graph: Graph, sinks: List[LineaID], keep_lineapy_save: bool = False
+) -> Graph:
     """
     Takes a full graph from the session
     and produces the subset responsible for the "sinks".
 
+    :param graph: A full graph objection from a session.
+    :param sinks: A list of node IDs desired for slicing.
+    :param keep_lineapy_save: Wheather to retain ``lineapy.save()`` in code slice.
+            Defaults to ``False``.
+    :return: A subgraph extracted (i.e., sliced) for the desired node IDs.
+
     """
+    if keep_lineapy_save:
+        # Children of an artifact sink include .save() statement.
+        # Identify .save() statement and make it the new sink.
+        # If not applicable, retain the original artifact sink.
+        new_sinks = []
+        for sink in sinks:
+            new_sink = sink
+            child_ids = graph.get_children(sink)
+            for c_id in child_ids:
+                c_node = graph.get_node(c_id)
+                if isinstance(c_node, CallNode) and c_node.source_location:
+                    source_code = c_node.source_location.source_code.code
+                    line_number = c_node.source_location.lineno
+                    line_code = source_code.split("\n")[line_number - 1]
+                    first_arg = c_node.positional_args[0]
+                    if "lineapy.save" in line_code and first_arg.id == sink:
```

Comment:
This regex is not great. A more robust implementation is to traverse the call nodes to check that a `getattr` call with `"save"` as the name was called on the result of some import functions on the name "lineapy". This will expand the scope. But I don't feel comfortable introducing regex into our core codebase.

Reply:
Yes, that makes sense! I can start looking into the direction you suggested.